pkg/bench/query_runner.go

```diff
@@ -0,0 +1,274 @@
+	config_util "github.com/prometheus/common/config"
+	"github.com/thanos-io/thanos/pkg/discovery/dns"
+	"github.com/thanos-io/thanos/pkg/extprom"
+)
+
+type QueryConfig struct {
+	Enabled           bool   `yaml:"enabled"`
+	Endpoint          string `yaml:"endpoint"`
+	BasicAuthUsername string `yaml:"basic_auth_username"`
+	BasicAuthPasword  string `yaml:"basic_auth_password"`
+}
+
+func (cfg *QueryConfig) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, "bench.query.enabled", false, "enable query benchmarking")
+	f.StringVar(&cfg.Endpoint, "bench.query.endpoint", "", "Remote query endpoint.")
+	f.StringVar(&cfg.BasicAuthUsername, "bench.query.basic-auth-username", "", "Set the basic auth username on remote query requests.")
+	f.StringVar(&cfg.BasicAuthPasword, "bench.query.basic-auth-password", "", "Set the basic auth password on remote query requests.")
+}
+
+type queryRunner struct {
+	id         string
+	tenantName string
+	cfg        QueryConfig
+
+	// Do DNS client side load balancing if configured
+	dnsProvider *dns.Provider
+	remoteMtx   sync.Mutex
+	addresses   []string
+	clientPool  map[string]v1.API
+
+	workload *queryWorkload
+
+	reg    prometheus.Registerer
+	logger log.Logger
+
+	requestDuration *prometheus.HistogramVec
+}
+
+func newQueryRunner(id string, tenantName string, cfg QueryConfig, workload *queryWorkload, logger log.Logger, reg prometheus.Registerer) (*queryRunner, error) {
+	runner := &queryRunner{
+		id:         id,
+		tenantName: tenantName,
+		cfg:        cfg,
+
+		workload:   workload,
+		clientPool: map[string]v1.API{},
+		dnsProvider: dns.NewProvider(
+			logger,
+			extprom.WrapRegistererWithPrefix("benchtool_", reg),
+			dns.GolangResolverType,
+		),
+
+		logger: logger,
+		reg:    reg,
+		requestDuration: promauto.With(reg).NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "benchtool",
+				Name:      "query_request_duration_seconds",
+				Buckets:   []float64{0.001, 0.01, 0.1, 0.3, 0.6, 1, 3, 6, 9, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 200},
+			},
+			[]string{"code", "type"},
+		),
+	}
+
+	return runner, nil
+}
+
+// jitterUp adds random jitter to the duration.
+//
+// This adds or subtracts time from the duration within a given jitter fraction.
+// For example for 10s and jitter 0.1, it will return a time within [9s, 11s])
+//
+// Reference: https://godoc.org/github.com/grpc-ecosystem/go-grpc-middleware/util/backoffutils
+func jitterUp(duration time.Duration, jitter float64) time.Duration {
+	multiplier := jitter * (rand.Float64()*2 - 1)
+	return time.Duration(float64(duration) * (1 + multiplier))
+}
+
+func (q *queryRunner) Run(ctx context.Context) error {
+	go q.resolveAddrsLoop(ctx)
+
+	queryChan := make(chan query, 1000)
+	for i := 0; i < 100; i++ {
+		go q.queryWorker(queryChan)
+	}
+	for _, queryReq := range q.workload.queries {
+		// every query has a ticker and a Go loop...
+		// not sure if this is a good idea but it should be fine
+		go func(q query) {
```

Comment:
Shadowing the receiver here makes it kinda hard to follow this.

Reply:
Good point. I'll change it to `req`